python/ray/autoscaler/docker.py

```diff
@@ -8,21 +8,61 @@
 logger = logging.getLogger(__name__)
 
 
-def dockerize_if_needed(config):
+def docker_pull_if_needed(config, commands):
+    if "docker" not in config:
+        return
+    if config["docker"].get("pull_before_run", True) is False:
+        return
+
+    docker_image = config["docker"].get("image")
+    docker_pull_cmd = "docker pull {}".format(docker_image)
+    commands.append(docker_pull_cmd)
+
+
+def dockerize_worker_setup_if_needed(config, commands):
     if "docker" not in config:
         return config
 
     docker_image = config["docker"].get("image")
-    docker_pull = config["docker"].get("pull_before_run", True)
+    ssh_user = config["auth"]["ssh_user"]
+    worker_docker_image = config["docker"].get("worker_image", docker_image)
+    docker_mounts = {dst: dst for dst in config["file_mounts"]}
     cname = config["docker"].get("container_name")
     run_options = config["docker"].get("run_options", [])
+    worker_run_options = config["docker"].get("worker_run_options", [])
+    worker_docker_start = docker_start_cmds(ssh_user, worker_docker_image,
+                                            docker_mounts, cname,
+                                            run_options + worker_run_options)
+    return worker_docker_start + (with_docker_exec(
+        commands, container_name=cname))
+
 
+def dockerize_head_setup_if_needed(config, commands):
+    docker_image = config["docker"].get("image")
+    ssh_user = config["auth"]["ssh_user"]
     head_docker_image = config["docker"].get("head_image", docker_image)
+    docker_mounts = {dst: dst for dst in config["file_mounts"]}
+    cname = config["docker"].get("container_name")
+    run_options = config["docker"].get("run_options", [])
     head_run_options = config["docker"].get("head_run_options", [])
+    head_docker_start = docker_start_cmds(ssh_user, head_docker_image,
+                                          docker_mounts, cname,
+                                          run_options + head_run_options)
 
-    worker_docker_image = config["docker"].get("worker_image", docker_image)
-    worker_run_options = config["docker"].get("worker_run_options", [])
+    return head_docker_start + (with_docker_exec(
+        config["head_setup_commands"], container_name=cname))
+
+
+def dockerize_if_needed(config):
```

Comment:
This is a pretty big change to this file not related to this PR. Please try to keep diffs minimal (revert these changes). I think you only need to:
1) Add a for loop in this function that rewrites the type-specific commands as worker docker commands.
2) Add a second for loop for the type-specific init commands.